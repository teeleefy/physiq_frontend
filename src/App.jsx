//import custom hook for accessing local storage state
import useLocalStorageState from './hooks/useLocalStorageState';
//react basic imports
import { useState, useEffect } from "react";
import './styles/App.css'
import {decodeToken} from "react-jwt";

//api and context import
import PhysiqApi from "./Api.js";
import UserContext from "./auth/UserContext";

//Route and NavBar imports
import NavBar from "./navigation/NavBar.jsx";
import AppRoutes from "./routes/Routes.jsx";
import Loading from "./navigation/Loading.jsx"

function App() {
  const [isLoading, setIsLoading] = useState(true);
  const [currentUser, setCurrentUser] =useState();
  const [familyMemberIds, setFamilyMemberIds] = useState([]);
  const [token, setToken] = useLocalStorageState("token");

    // Load user info from API. Until a user is logged in and they have a token,
  // this should not run. It only needs to re-run when a user logs out, so
  // the value of the token is a dependency for this effect.

  useEffect(function loadUserInfo() {
    console.debug("App useEffect loadUserInfo", "token=", token);

    async function getCurrentUser() {
      if (token) {
        try {
          let { familyId, memberIds } = decodeToken(token);
          // put the token on the Api class so it can use it to call the API.
          PhysiqApi.token = token;
          let currentUser = await PhysiqApi.getCurrentFamily(familyId);
          let familyMemberIds = new Set(memberIds);
          setCurrentUser(currentUser);
          setFamilyMemberIds(familyMemberIds);
          
        } catch (err) {
          console.error("App loadUserInfo: problem loading", err);
          setCurrentUser(null);
        }
        // setIsLoading(false);
      }
    }
    // set infoLoaded to false while async getCurrentUser runs; once the
    // data is fetched (or even if an error happens!), this will be set back
    // to false to control the spinner.
    setIsLoading(false);
    getCurrentUser();
  }, [token]);


  /** Handles site-wide login.
   *
   * Make sure you await this function and check its return value!
   */
  async function login(loginData) {
    try {
      let token = await PhysiqApi.loginUser(loginData);
      setToken(token);
      return { success: true };
    } catch (errors) {
      console.error("login failed", errors);
      return { success: false, errors };
    }
  }

  async function updateUser(updatedUserData) {
    try {
      let updatedUser = await PhysiqApi.updateCurrentFamily(currentUser.id, updatedUserData);
      setCurrentUser(updatedUser);
      return { success: true };
    } catch (errors) {
      console.error("Save failed", errors);
      return { success: false, errors };
    }
  }


  /** Handles site-wide logout. */
  function logout() {
    setCurrentUser(null);
    setToken(null);
  }

   /** Handles site-wide signup.
   *
   * Automatically logs them in (set token) upon signup.
   *
   * Make sure you await this function and check its return value!
   */
   async function signup(signupData) {
    try {
      let token = await PhysiqApi.signupUser(signupData);
      setToken(token);
      return { success: true };
    } catch (errors) {
      console.error("signup failed", errors);
      return { success: false, errors };
    }
  }



  if (isLoading) return <Loading/>;

  return (
    <UserContext.Provider value={{currentUser, token, setCurrentUser}}>
      <div className="App">
          <NavBar logout={logout}/>
          <main>
            <AppRoutes login={login} signup={signup} update={updateUser}/>
          </main>
      </div>
    </UserContext.Provider>  
  )
}

export default App
