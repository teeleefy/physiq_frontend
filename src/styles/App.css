main {
  width: 100vw;
  height: 95vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  padding-top: 0vh;
  color: black;
}

#App-main{
  padding-top: 2rem;
}

.card {
  border-radius: 0;
  box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.14),
    0 3px 1px -2px rgba(0, 0, 0, 0.12), 0 1px 5px 0 rgba(0, 0, 0, 0.2);
  margin-bottom: 15vh;
}


.mychart {
  color: red;
}

#main-navbar{
  height: 7vh;
}

/* #footer{
  position: fixed;
  height: 5vh;
  bottom: 0;
  width: 100%;
} */

#footer{
  position: absolute;
  height: 5vh;
  
  margin:auto;
  text-align: center;
  bottom: 0.5rem;
  width: 100%;
}

