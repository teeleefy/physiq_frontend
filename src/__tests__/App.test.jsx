import { renderWithRouter, screen } from '../../__tests__/test-tools';
import App from '../App'

describe('App', () => {
  it('renders the App component', () => {
    renderWithRouter(<App />)
    screen.debug(); // prints out the jsx in the App component unto the command line
  })

  it("matches snapshot", function() {
    const { asFragment } = renderWithRouter(<App />);
    expect(asFragment()).toMatchSnapshot();
  });
})




