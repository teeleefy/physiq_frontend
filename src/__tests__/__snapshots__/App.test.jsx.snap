// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div
      id="main-navbar"
    >
      <div>
        <nav
          class="navbar navbar-expand-md"
        >
          <div
            class="container-fluid"
          >
            <a
              aria-current="page"
              class="navbar-brand active"
              href="/"
            >
              fa
              <b
                class="mychart"
              >
                m
              </b>
              il
              <b
                class="mychart"
              >
                y
              </b>
               
              <b
                class="mychart"
              >
                chart
              </b>
            </a>
            <ul
              class="ml-auto navbar-nav"
            >
              <li
                class="nav-item"
              >
                <a
                  class=""
                  href="/login"
                >
                  Login
                </a>
                <a
                  class=""
                  href="/signup"
                >
                  Sign Up
                </a>
              </li>
            </ul>
          </div>
        </nav>
      </div>
    </div>
    <main
      id="App-main"
    >
      <div>
        <div
          id="FamilyChart-main"
        >
          <div
            id="FamilyChart-message"
          >
            <div
              class="Home-btns"
            >
              <div
                class="text-center card-title"
              >
                <h1
                  class="Home-title"
                >
                  Welcome to fa
                  <b
                    class="mychart"
                  >
                    m
                  </b>
                  il
                  <b
                    class="mychart"
                  >
                    y
                  </b>
                   
                  <b
                    class="mychart"
                  >
                    chart
                  </b>
                  !
                </h1>
              </div>
              <p
                class="Home-text card-text"
              >
                Keep track of your family's health, all in one, convenient place!
              </p>
              <a
                class=""
                href="/login"
              >
                <button
                  class="FamilyChart-btn btn btn-secondary"
                >
                  Login
                </button>
              </a>
              <a
                class=""
                href="/signup"
              >
                <button
                  class="FamilyChart-btn justify-content-center btn btn-secondary"
                >
                  Sign Up
                </button>
              </a>
            </div>
          </div>
        </div>
      </div>
    </main>
    <div
      id="footer"
    >
      <div>
        <a
          class="Footer-navlink"
          href="https://tammypainter.carrd.co/"
          id="Footer-linkedin"
        >
          <i
            class="fa-solid fa-user"
          />
        </a>
        <a
          class="Footer-navlink"
          href="https://www.linkedin.com/in/tammy-painter-73416b282/"
          id="Footer-linkedin"
        >
          <i
            class="fa-brands fa-linkedin"
          />
        </a>
        <a
          class="Footer-navlink"
          href="https://github.com/teeleefy"
          id="Footer-github"
        >
          <i
            class="fa-brands fa-github"
          />
        </a>
        <a
          class="Footer-navlink"
          href="mailto:tammypainter.codes@gmail.com"
          id="Footer-envelope"
        >
          <i
            class="fa-regular fa-envelope"
          />
        </a>
        <a
          aria-current="page"
          class="active"
          href="/"
          id="Footer-copyright"
        >
          Ⓒ 
          <span
            id="Footer-rights"
          >
             2024   family chart    All rights reserved
          </span>
        </a>
      </div>
    </div>
    <div
      class="Toastify"
    />
  </div>
</DocumentFragment>
`;
